File: src_java/main/java/com/github/davinkevin/transmissionrss/batch/synchronization/feeds/specification/FeedsSpecificationReader.java
package com.github.davinkevin.transmissionrss.batch.synchronization.feeds.specification;

import com.github.davinkevin.transmissionrss.batch.synchronization.feeds.specification.model.FeedSpecification;
import com.github.davinkevin.transmissionrss.feeds.properties.FeedsProperty;
import io.vavr.Tuple2;
import io.vavr.collection.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.batch.item.ItemReader;
import org.springframework.stereotype.Component;

import static io.vavr.API.Tuple;

/**
 * Created by kevin on 06/05/2018
 */
@Slf4j
@Component
public class FeedsSpecificationReader implements ItemReader<FeedSpecification> {

    private List<FeedSpecification> pm;

    public FeedsSpecificationReader(FeedsProperty feedsProperty) {
        pm = feedsProperty.getFeeds()
                .map(f -> Tuple(f, f.getRegexp()))
                .flatMap(t -> t._2().map(v -> FeedSpecification.builder()
                        .url(t._1().getUrl())
                        .downloadPath(v.getDownloadPath())
                        .exclude(v.getExclude())
                        .matcher(v.getMatcher())
                        .build())
                )
        ;
    }

    @Override
    public FeedSpecification read() {
        if (pm.isEmpty()) {
            return null;
        }

        Tuple2<FeedSpecification, List<FeedSpecification>> p = pm.pop2();
        pm = p._2();

        log.info("Push value {}", p._1());

        return p._1();
    }
}

File: src/main/kotlin/com/github/davinkevin/transmissionrss/batch/synchronization/feeds/specification/FeedsSpecificationReader.kt
package com.github.davinkevin.transmissionrss.batch.synchronization.feeds.specification

import com.github.davinkevin.transmissionrss.feeds.properties.FeedsProperty
import org.springframework.batch.item.ItemReader
import org.springframework.stereotype.Component

@Component
class FeedsSpecificationReader(feedsProperty: FeedsProperty) : ItemReader<FeedSpecification> {

    lateinit var pm: List<FeedSpecification>

    init {
        pm = feedsProperty
                .feeds
                .map { it to it.regexp }
                .flatMap { v -> v.second
                        .map { r ->
                            FeedSpecification(
                                    url = v.first.url!!,
                                    matcher = r.matcher!!,
                                    exclude = r.exclude,
                                    downloadPath = r.downloadPath!!
                            )
                        }
                }
    }

    override fun read(): FeedSpecification? {
        val p = pm.firstOrNull()
        pm = pm.filter { it != p }

        return p
    }
}
